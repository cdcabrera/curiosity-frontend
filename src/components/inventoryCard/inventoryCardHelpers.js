import React from 'react';
import { cellWidth as PfCellWidth, SortByDirection, wrappable } from '@patternfly/react-table';
import _camelCase from 'lodash/camelCase';
import { Tooltip } from '../tooltip/tooltip';
import { translate } from '../i18n/i18n';
import {
  RHSM_API_QUERY_SORT_DIRECTION_TYPES as SORT_DIRECTION_TYPES,
  RHSM_API_QUERY_TYPES
} from '../../types/rhsmApiTypes';

// ToDo: review setting up a transformed cell cache for already transformed cells.

/**
 * Apply product inventory config properties consistently.
 *
 * @param {Function|string|number} prop
 * @param {object} options
 * @param {Array|object} options.params
 * @returns {Node|string|number|undefined}
 */
const applyConfigProperty = (prop, { params = [] } = {}) => {
  let updatedProp = prop;

  if (typeof prop === 'function') {
    updatedProp = prop(...((Array.isArray(params) && params) || [params]));
  }

  if (typeof updatedProp === 'string' || typeof updatedProp === 'number' || React.isValidElement(updatedProp)) {
    return updatedProp;
  }

  return undefined;
};

/**
 * Apply header and row cell product filters.
 *
 * @param {Array} filters
 * @param {object} cellData
 * @param {object} session
 * @returns {{bodyCells: Array, headerCells: Array}}
 */
const applyHeaderRowCellFilters = (filters = [], cellData = {}, session = {}) => {
  const headerCells = [];
  const bodyCells = [];

  filters.forEach(({ id, cell, cellWidth, header, onSort, sortId, sortActive, sortDirection, tooltip, transforms }) => {
    const headerCellUpdated = { title: translate('curiosity-inventory.header', { context: id }) };
    const bodyCellUpdated = { title: '' };

    // set filtered base header and body cells, or if filter doesn't exist skip
    if (cellData[id]) {
      headerCellUpdated.title = cellData[id]?.title ?? id;
      bodyCellUpdated.title = cellData[id]?.value ?? '';
    } else {
      return;
    }

    // set header cell title
    if (header) {
      const updatedHeaderCellTitle = applyConfigProperty(header, { params: cellData });
      if (updatedHeaderCellTitle) {
        headerCellUpdated.title = updatedHeaderCellTitle;
      }
    }

    // set header cell tooltip
    if (tooltip) {
      const updatedTooltip = applyConfigProperty(tooltip, { params: cellData });
      if (updatedTooltip) {
        headerCellUpdated.title = <Tooltip content={updatedTooltip}>hey</Tooltip>;
      }
    }

    // set header cell transforms
    if (headerCellUpdated) {
      headerCellUpdated.transforms = [];

      if (Array.isArray(transforms)) {
        headerCellUpdated.transforms = headerCellUpdated.transforms.concat([...transforms]);
      }

      if (typeof cellWidth === 'number') {
        headerCellUpdated.transforms.push(PfCellWidth(cellWidth));
      }
    }

    // set header cell onSort
    if (typeof onSort === 'function') {
      headerCellUpdated.onSort = obj => onSort({ ...cellData }, { ...obj, id: sortId || id });
      headerCellUpdated.sortActive = sortActive;
      headerCellUpdated.sortDirection = sortDirection;
    }

    // set body cell title
    if (cell) {
      const updatedBodyCellTitle = applyConfigProperty(cell, { params: [{ ...cellData }, { ...session }] });
      if (updatedBodyCellTitle) {
        bodyCellUpdated.title = updatedBodyCellTitle;
      }
    }

    headerCells.push(headerCellUpdated);
    bodyCells.push(bodyCellUpdated);
  });

  return {
    headerCells,
    bodyCells
  };
};

/**
 * Apply sort filter to filters.
 *
 * @param {object} params
 * @param {{ onSort: Function, sortActive: boolean, sortDirection: string, isSortDefault: boolean,
 *     sortDefaultInitialDirection: string }} params.filter
 * @param {Function} params.onSort
 * @param {object} params.query
 * @returns {object}
 */
const applySortFilters = ({ filter = {}, onSort, query = {} }) => {
  const { id, sortId } = filter;
  const updatedId = sortId || id;
  const updatedFilter = { ...filter };
  const hasSort = updatedFilter.onSort || onSort;

  if (!updatedFilter.onSort && onSort) {
    updatedFilter.onSort = onSort;
  }

  // set fallback for the active sorted column based on query
  if (
    hasSort &&
    typeof updatedFilter.sortActive !== 'boolean' &&
    query?.[RHSM_API_QUERY_TYPES.SORT] &&
    (query?.[RHSM_API_QUERY_TYPES.SORT] === updatedId || _camelCase(query?.[RHSM_API_QUERY_TYPES.SORT]) === updatedId)
  ) {
    updatedFilter.sortActive = true;
  }

  // set sort direction
  if (hasSort && !updatedFilter.sortDirection && query?.[RHSM_API_QUERY_TYPES.DIRECTION]) {
    switch (query?.[RHSM_API_QUERY_TYPES.DIRECTION]) {
      case SORT_DIRECTION_TYPES.DESCENDING:
        updatedFilter.sortDirection = SortByDirection.desc;
        break;
      default:
        updatedFilter.sortDirection = SortByDirection.asc;
        break;
    }
  }

  if (
    hasSort &&
    !updatedFilter.sortActive &&
    !query?.[RHSM_API_QUERY_TYPES.SORT] &&
    updatedFilter.isSortDefault === true
  ) {
    updatedFilter.sortActive = true;

    if (updatedFilter.sortDefaultInitialDirection) {
      updatedFilter.sortDirection = updatedFilter.sortDefaultInitialDirection;
    }
  }

  return updatedFilter;
};

const applyWrappableFilters = ({ filter = {} }) => {
  const updatedFilter = { ...filter };

  if (Array.isArray(updatedFilter.transforms)) {
    updatedFilter.transforms.push(wrappable);
  } else {
    updatedFilter.transforms = [wrappable];
  }

  return updatedFilter;
};

/**
 * Apply additional properties to filters.
 *
 * @param {object} params
 * @param {Array} params.filters
 * @param {Function} params.onSort
 * @param {object} params.query
 * @returns {Array}
 */
const parseInventoryFilters = ({ filters = [], onSort, query = {} }) =>
  [...filters].map(filter => {
    const updatedFilter = { ...filter };

    if (updatedFilter.isSortable) {
      Object.assign(updatedFilter, applySortFilters({ filter: updatedFilter, onSort, query }));
    }

    if (updatedFilter.isWrappable) {
      Object.assign(updatedFilter, applyWrappableFilters({ filter: updatedFilter }));
    }

    return updatedFilter;
  });

/**
 * Parse and return formatted/filtered table cells, and apply table filters.
 *
 * @param {object} params
 * @param {Array} params.filters
 * @param {object} params.cellData
 * @param {object} params.session
 * @returns {{columnHeaders: Array, cells: Array, data: object}}
 */
const parseRowCellsListData = ({ filters = [], cellData = {}, session = {} }) => {
  const updatedColumnHeaders = [];
  const updatedCells = [];
  const allCells = {};

  // Apply basic translation and value
  Object.entries(cellData).forEach(([key, value = '']) => {
    allCells[key] = {
      title: translate('curiosity-inventory.header', { context: key }),
      value
    };

    updatedColumnHeaders.push(allCells[key].title);
    updatedCells.push(value || '...');
  });

  // Apply filters to header and cell values
  if (filters?.length && Object.keys(allCells).length) {
    updatedColumnHeaders.length = 0;
    updatedCells.length = 0;

    const { headerCells = [], bodyCells = [] } = applyHeaderRowCellFilters(filters, allCells, session);

    updatedColumnHeaders.push(...headerCells);
    updatedCells.push(...bodyCells);
  }

  return {
    columnHeaders: updatedColumnHeaders,
    cells: updatedCells,
    data: { ...allCells }
  };
};

const inventoryCardHelpers = {
  applyConfigProperty,
  applyHeaderRowCellFilters,
  applySortFilters,
  applyWrappableFilters,
  parseInventoryFilters,
  parseRowCellsListData
};

export {
  inventoryCardHelpers as default,
  inventoryCardHelpers,
  applyConfigProperty,
  applyHeaderRowCellFilters,
  applySortFilters,
  applyWrappableFilters,
  parseInventoryFilters,
  parseRowCellsListData
};
