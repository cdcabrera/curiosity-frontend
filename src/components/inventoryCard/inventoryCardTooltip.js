import React from 'react';
import PropTypes from 'prop-types';
import { Tooltip as PfTooltip, TooltipPosition } from '@patternfly/react-core';

/**
 * PF tooltip wrapper component.
 *
 * @param {object} props
 * @param {Node} props.children
 * @param {Node} props.content
 * @returns {Node}
 */
const Tooltip = ({ children, content, ...props }) => (
  <PfTooltip content={(React.isValidElement(content) && content) || <p>{content || ''}</p>} {...props}>
    {(React.isValidElement(children) && children) || <span className="curiosity-tooltip-children">{children}</span>}
  </PfTooltip>
);

/**
 * Prop types.
 *
 * @type {{children: Node, content: Node}}
 */
Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  content: PropTypes.node,
  distance: PropTypes.number,
  enableFlip: PropTypes.bool,
  entryDelay: PropTypes.number,
  exitDelay: PropTypes.number,
  position: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{content: string}}
 */
Tooltip.defaultProps = {
  content: '...',
  distance: 5,
  enableFlip: false,
  entryDelay: 100,
  exitDelay: 0,
  position: TooltipPosition.top
};

export { Tooltip as default, Tooltip };
