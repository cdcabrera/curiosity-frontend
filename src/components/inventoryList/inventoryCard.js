import React from 'react';
import PropTypes from 'prop-types';
import { Bullseye, Card, CardActions, CardBody, CardFooter, CardHeader, CardHeaderMain } from '@patternfly/react-core';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import { MinHeight } from '../minHeight/minHeight';
import Pagination from '../pagination/pagination';
import { ToolbarFieldDisplayName } from '../toolbar/toolbarFieldDisplayName';
import { paginationHelpers } from '../pagination/paginationHelpers';
import { translate } from '../i18n/i18n';

/**
 * A hosts system inventory component.
 *
 * @fires onColumnSort
 * @fires onPage
 * @param {object} props
 * @param {boolean} props.isDisabled
 * @param {boolean} props.error
 * @param {boolean} props.fulfilled
 * @param {number} props.itemCount
 * @param {boolean} props.pending
 * @param {Function} props.t
 * @returns {Node}
 */
const InventoryList = ({
  isDisabled,
  error,
  fulfilled,
  itemCount,
  pending,
  t,
  updatedOffset,
  updatedPerPage,
  onPage,
  children
}) => {
  if (isDisabled) {
    return (
      <Card className="curiosity-inventory-card__disabled">
        <CardBody>
          <Bullseye>{t('curiosity-inventory.tab', { context: 'disabled' })}</Bullseye>
        </CardBody>
      </Card>
    );
  }

  const isLastPage = paginationHelpers.isLastPage(updatedOffset, updatedPerPage, itemCount);

  // Set an updated key to force refresh minHeight
  const minHeightContentRefreshKey =
    (fulfilled === true && itemCount < updatedPerPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (fulfilled === true && isLastPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (error === true && `bodyMinHeight-${updatedPerPage}-resize`) ||
    `bodyMinHeight-${updatedPerPage}`;

  return (
    <Card className="curiosity-inventory-card">
      <MinHeight key="headerMinHeight" updateOnContent>
        <CardHeader className={(error && 'hidden') || ''} aria-hidden={error || false}>
          <CardHeaderMain>
            <ToolbarFieldDisplayName />
          </CardHeaderMain>
          <CardActions className={(!itemCount && 'transparent') || ''} aria-hidden={!itemCount || false}>
            <Pagination
              isCompact
              isDisabled={pending || error}
              itemCount={itemCount}
              offset={updatedOffset}
              onPage={onPage}
              onPerPage={onPage}
              perPage={updatedPerPage}
            />
          </CardActions>
        </CardHeader>
      </MinHeight>
      <MinHeight key={minHeightContentRefreshKey} updateOnContent>
        <CardBody>
          <div className={(error && 'blur') || 'fadein'}>{children}</div>
        </CardBody>
      </MinHeight>
      <MinHeight key="footerMinHeight" updateOnContent>
        <CardFooter
          className={(error && 'hidden') || (!itemCount && 'transparent') || ''}
          aria-hidden={error || !itemCount || false}
        >
          <Pagination
            dropDirection="up"
            isDisabled={pending || error}
            itemCount={itemCount}
            offset={updatedOffset}
            onPage={onPage}
            onPerPage={onPage}
            perPage={updatedPerPage}
          />
        </CardFooter>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{t: Function, listData: Array, session: object, pending: boolean, fulfilled: boolean,
 *     getHostsInventory: Function, perPageDefault: number, isDisabled: boolean, error: boolean,
 *     itemCount: number}}
 */
InventoryList.propTypes = {
  error: PropTypes.bool,
  fulfilled: PropTypes.bool,
  isDisabled: PropTypes.bool,
  itemCount: PropTypes.number,
  pending: PropTypes.bool,
  t: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{t: Function, listData: Array, session: object, pending: boolean, fulfilled: boolean,
 *     getHostsInventory: Function, perPageDefault: number, isDisabled: boolean, error: boolean,
 *     itemCount: number}}
 */
InventoryList.defaultProps = {
  error: false,
  fulfilled: false,
  isDisabled: false,
  itemCount: 0,
  pending: false,
  t: translate
};

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.inventoryList.makeInventoryList();

const ConnectedInventoryList = connect(makeMapStateToProps, mapDispatchToProps)(InventoryList);

export { ConnectedInventoryList as default, ConnectedInventoryList, InventoryList };
