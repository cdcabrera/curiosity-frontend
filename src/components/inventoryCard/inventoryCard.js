import React from 'react';
import PropTypes from 'prop-types';
import { Bullseye, Card, CardActions, CardBody, CardFooter, CardHeader, CardHeaderMain } from '@patternfly/react-core';
import { TableVariant } from '@patternfly/react-table';
import { TableToolbar } from '@redhat-cloud-services/frontend-components/TableToolbar';
import {
  useProduct,
  useProductInventoryInstancesQuery,
  useProductInventoryInstancesConfig
} from '../productView/productViewContext';
import { RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { paginationHelpers } from '../pagination/paginationHelpers';
import { MinHeight } from '../minHeight/minHeight';
import { ToolbarFieldDisplayName } from '../toolbar/toolbarFieldDisplayName';
import { Pagination } from '../pagination/pagination';
import { Loader } from '../loader/loader';
import { helpers } from '../../common/helpers';
import { translate } from '../i18n/i18n';
import Table from '../table/table';
import {inventoryListHelpers} from "../inventoryList/inventoryListHelpers";
import GuestsList from "../guestsList/guestsList";

/**
 * An inventory component.
 *
 * @param root0
 * @param root0.useProduct
 * @param root0.useProductInventoryInstancesConfig
 * @param root0.useProductInventoryInstancesQuery
 * @param root0.isDisabled
 * @param root0.t
 * @param root0.useInventorySelector
 * @returns {null}
 */
const InventoryList = ({
  isDisabled,
  t,
  useInventorySelector: useAliasInventorySelector,
  useProduct: useAliasProduct,
  useProductInventoryInstancesConfig: useAliasProductInventoryInstancesConfig,
  useProductInventoryInstancesQuery: useAliasProductInventoryInstancesQuery
}) => {
  const { viewId } = useAliasProduct();
  const query = useAliasProductInventoryInstancesQuery();
  const { settings, filters: filterInventoryData } = useAliasProductInventoryInstancesConfig();
  const session = useAliasSessionSelector();
  const { error, fulfilled, pending, listData = [] } = useAliasInventorySelector();

  /*
  const {
    error,
    filterInventoryData,
    fulfilled,
    isDisabled,
    itemCount,
    listData,
    pending,
    perPageDefault,
    t
  }
  */

  if (isDisabled) {
    return (
      <Card className="curiosity-inventory-card__disabled">
        <CardBody>
          <Bullseye>{t('curiosity-inventory.tab', { context: 'disabled' })}</Bullseye>
        </CardBody>
      </Card>
    );
  }

  const onColumnSort = () => {};

  const onPage = () => {};

  const itemCount = 0;
  const updatedPerPage = query[RHSM_API_QUERY_TYPES.LIMIT] || 10;
  const updatedOffset = query[RHSM_API_QUERY_TYPES.OFFSET];
  const isLastPage = paginationHelpers.isLastPage(updatedOffset, updatedPerPage, itemCount);

  // Set an updated key to force refresh minHeight
  const minHeightContentRefreshKey =
    (fulfilled === true && itemCount < updatedPerPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (fulfilled === true && isLastPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (error === true && `bodyMinHeight-${updatedPerPage}-resize`) ||
    `bodyMinHeight-${updatedPerPage}`;

  let updatedColumnHeaders = [];

  const updatedRows = listData.map(({ ...cellData }) => {
    const { columnHeaders, cells } = inventoryListHelpers.parseRowCellsListData({
      filters: inventoryListHelpers.parseInventoryFilters({
        filters: filterInventoryData,
        onSort: onColumnSort,
        query
      }),
      cellData,
      session
    });

    updatedColumnHeaders = columnHeaders;

    const guestsId = cellData?.subscriptionManagerId;
    let hasGuests = cellData?.numberOfGuests > 0 && guestsId;

    // Apply hasGuests callback, return boolean
    if (typeof settings?.hasGuests === 'function') {
      hasGuests = settings.hasGuests({ ...cellData }, { ...session });
    }

    return {
      cells,
      expandedContent:
        (hasGuests && (
          <GuestsList
            key={guestsId}
            filterGuestsData={filterGuestsData}
            numberOfGuests={cellData?.numberOfGuests}
            id={guestsId}
            query={query}
          />
        )) ||
        undefined
    };
  });

  return (
    <Card className="curiosity-inventory-card">
      <MinHeight key="headerMinHeight" updateOnContent>
        <CardHeader className={(error && 'hidden') || ''} aria-hidden={error || false}>
          <CardHeaderMain>
            <ToolbarFieldDisplayName viewId={viewId} />
          </CardHeaderMain>
          <CardActions className={(!itemCount && 'transparent') || ''} aria-hidden={!itemCount || false}>
            <Pagination
              isCompact
              isDisabled={pending || error}
              itemCount={itemCount}
              offset={updatedOffset}
              onPage={onPage}
              onPerPage={onPage}
              perPage={updatedPerPage}
            />
          </CardActions>
        </CardHeader>
      </MinHeight>
      <MinHeight key={minHeightContentRefreshKey} updateOnContent>
        <CardBody>
          <div className={(error && 'blur') || (pending && 'fadein') || ''}>
            {pending && (
              <Loader
                variant="table"
                tableProps={{
                  className: 'curiosity-inventory-list',
                  colCount: filterInventoryData?.length || (listData?.[0] && Object.keys(listData[0]).length) || 1,
                  colWidth:
                    (filterInventoryData?.length && filterInventoryData.map(({ cellWidth }) => cellWidth)) || [],
                  rowCount: listData?.length || updatedPerPage,
                  variant: TableVariant.compact
                }}
              />
            )}
            {!pending && (
              <Table
                borders
                variant={TableVariant.compact}
                className="curiosity-inventory-list"
                columnHeaders={updatedColumnHeaders}
                rows={updatedRows}
              />
            )}
          </div>
        </CardBody>
      </MinHeight>
      <MinHeight key="footerMinHeight" updateOnContent>
        <CardFooter
          className={(error && 'hidden') || (!itemCount && 'transparent') || ''}
          aria-hidden={error || !itemCount || false}
        >
          <TableToolbar isFooter>
            <Pagination
              dropDirection="up"
              isDisabled={pending || error}
              itemCount={itemCount}
              offset={updatedOffset}
              onPage={onPage}
              onPerPage={onPage}
              perPage={updatedPerPage}
            />
          </TableToolbar>
        </CardFooter>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{}}
 */
InventoryList.propTypes = {
  isDisabled: PropTypes.bool,
  t: PropTypes.func,
  useInventorySelector: PropTypes.func,
  useProduct: PropTypes.func,
  useProductInventoryInstancesQuery: PropTypes.func,
  useProductInventoryInstancesConfig: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{}}
 */
InventoryList.defaultProps = {
  isDisabled: helpers.UI_DISABLED_TABLE_HOSTS,
  t: translate,
  useInventorySelector,
  useProduct,
  useProductInventoryInstancesConfig,
  useProductInventoryInstancesQuery
};

export { InventoryList as default, InventoryList };
