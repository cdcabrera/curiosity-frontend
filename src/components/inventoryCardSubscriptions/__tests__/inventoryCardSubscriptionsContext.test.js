import {
  context,
  useGetSubscriptionsInventory,
  useInventoryCardActionsSubscriptions,
  useOnPageSubscriptions,
  useOnColumnSortSubscriptions,
  useParseSubscriptionsFiltersSettings,
  useSelectorSubscriptions
} from '../inventoryCardSubscriptionsContext';
import {
  RHSM_API_QUERY_INVENTORY_SORT_DIRECTION_TYPES as SORT_DIRECTION_TYPES,
  RHSM_API_QUERY_INVENTORY_SUBSCRIPTIONS_SORT_TYPES as SORT_TYPES,
  RHSM_API_QUERY_SET_TYPES,
  RHSM_API_RESPONSE_SUBSCRIPTIONS_DATA_TYPES as SUBSCRIPTIONS_INVENTORY_TYPES
} from '../../../services/rhsm/rhsmConstants';
import { inventoryCardHelpers } from '../../inventoryCard/inventoryCardHelpers';

describe('InventoryCardSubscriptionsContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should expect specific sort properties', () => {
    expect(SORT_TYPES).toMatchSnapshot('sort properties');
  });

  /*
  it('should handle a store response with useGetSubscriptionsInventory', async () => {
    const { result } = await renderHook(
      () =>
        useGetSubscriptionsInventory({
          getInventory: () => () => {},
          useProduct: () => ({ productId: 'lorem' }),
          useDispatch: () => {}
        }),
      {
        state: {
          inventory: {
            subscriptionsInventory: {
              lorem: {
                fulfilled: true,
                data: [{ data: [{ lorem: 'ipsum' }, { dolor: 'sit' }], meta: {} }]
              }
            }
          }
        }
      }
    );

    expect(result).toMatchSnapshot('store response');
  });
  */

  it('should handle variations in instances inventory API responses', async () => {
    const { result: errorResponse } = await renderHook(() =>
      useGetSubscriptionsInventory({
        getInventory: () => () => {},
        useDispatch: () => {},
        useProduct: () => ({ productId: 'lorem' }),
        useProductInventoryQuery: () => ({}),
        useSelector: () => ({ error: true })
      })
    );

    expect(errorResponse).toMatchSnapshot('inventory, error');

    const { result: pendingResponse } = await renderHook(() =>
      useGetSubscriptionsInventory({
        getInventory: () => () => {},
        useDispatch: () => {},
        useProduct: () => ({ productId: 'lorem' }),
        useProductInventoryQuery: () => ({}),
        useSelector: () => ({ pending: true })
      })
    );

    expect(pendingResponse).toMatchSnapshot('inventory, pending');

    const { result: cancelledResponse } = await renderHook(() =>
      useGetSubscriptionsInventory({
        useDispatch: () => {},
        getInventory: () => () => {},
        useProduct: () => ({ productId: 'lorem' }),
        useProductInventoryQuery: () => ({}),
        useSelector: () => ({ cancelled: true })
      })
    );

    expect(cancelledResponse).toMatchSnapshot('inventory, cancelled');

    const { result: fulfilledResponse } = await renderHook(() =>
      useGetSubscriptionsInventory({
        useDispatch: () => {},
        getInventory: () => () => {},
        useProduct: () => ({ productId: 'lorem' }),
        useProductInventoryQuery: () => ({}),
        useSelector: () => ({ fulfilled: true })
      })
    );

    expect(fulfilledResponse).toMatchSnapshot('inventory, fulfilled');

    const { result: disabledResponse } = await renderHook(() =>
      useGetSubscriptionsInventory({
        isDisabled: true,
        useDispatch: () => {},
        getInventory: () => () => {},
        useProduct: () => ({ productId: 'lorem' }),
        useProductInventoryQuery: () => ({}),
        useSelector: () => ({ data: {}, fulfilled: false, pending: false, error: false })
      })
    );

    expect(disabledResponse).toMatchSnapshot('inventory, disabled');
  });

  it('should apply custom actions', async () => {
    const { result } = await renderHook(() =>
      useInventoryCardActionsSubscriptions({
        useSelector: () => ({
          resultsCount: 1
        }),
        useProductConfig: () => ({
          settings: {
            actions: [
              {
                id: RHSM_API_QUERY_SET_TYPES.DISPLAY_NAME
              },
              {
                content: () => 'hello world'
              }
            ]
          }
        })
      })
    );

    expect(result).toMatchSnapshot('custom actions');
  });

  it('should handle an onPage event', () => {
    const mockDispatch = jest.fn();
    const onPage = useOnPageSubscriptions({
      useDispatch: () => mockDispatch,
      useProduct: () => ({ productId: 'lorem' })
    });

    onPage({ offset: 1, perPage: 5 });
    expect(mockDispatch.mock.calls).toMatchSnapshot('onPage event, dispatch');
    mockDispatch.mockClear();
  });

  it('should handle an onColumnSort event', () => {
    const mockDispatch = jest.fn();
    const onColumnSort = useOnColumnSortSubscriptions({
      sortColumns: { LOREM_IPSUM_COLUMN_ONE: 'loremIpsumColumnOne' },
      useDispatch: () => mockDispatch,
      useProduct: () => ({ productId: 'lorem' })
    });

    onColumnSort({ direction: SORT_DIRECTION_TYPES.DESCENDING, data: { metric: 'loremIpsumColumnOne' } });
    onColumnSort({ direction: SORT_DIRECTION_TYPES.ASCENDING, data: { metric: 'loremIpsumColumnOne' } });

    expect(mockDispatch.mock.calls).toMatchSnapshot('onColumnSort event, dispatch');
    mockDispatch.mockClear();
  });

  it('should parse filters and settings', async () => {
    const { result } = await renderHook(() =>
      useParseSubscriptionsFiltersSettings({
        useProductConfig: () => ({
          filters: [
            {
              metric: 'lorem ipsum'
            }
          ]
        })
      })
    );

    expect(result).toMatchSnapshot('parsed');
  });

  it('should handle a store response using selectors', async () => {
    const { result } = await renderHook(
      () =>
        useSelectorSubscriptions({
          useParseFiltersSettings: () =>
            inventoryCardHelpers.normalizeInventorySettings({
              filters: [
                {
                  metric: SUBSCRIPTIONS_INVENTORY_TYPES.PRODUCT_NAME
                }
              ],
              settings: {},
              productId: 'lorem'
            }),
          useProduct: () => ({ productId: 'lorem' })
        }),
      {
        state: {
          inventory: {
            subscriptionsInventory: {
              lorem: {
                fulfilled: true,
                data: [
                  {
                    data: [
                      { [SUBSCRIPTIONS_INVENTORY_TYPES.PRODUCT_NAME]: 'lorem-ipsum' },
                      { [SUBSCRIPTIONS_INVENTORY_TYPES.PRODUCT_NAME]: 'dolor-sit' }
                    ],
                    meta: {}
                  }
                ]
              }
            }
          }
        }
      }
    );

    expect(result).toMatchSnapshot('store response');
  });
});
