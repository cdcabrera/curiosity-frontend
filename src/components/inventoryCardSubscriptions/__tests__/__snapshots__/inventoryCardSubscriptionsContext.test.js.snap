// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryCardSubscriptionsContext should apply custom actions: custom actions 1`] = `
[
  <ToolbarItem>
    <DisplayName
      isFilter={false}
    />
  </ToolbarItem>,
  <ToolbarItem>
    hello world
  </ToolbarItem>,
]
`;

exports[`InventoryCardSubscriptionsContext should expect specific sort properties: sort properties 1`] = `
{
  "NEXT_EVENT_DATE": "next_event_date",
  "NEXT_EVENT_TYPE": "next_event_type",
  "PRODUCT_NAME": "product_name",
  "QUANTITY": "quantity",
  "SERVICE_LEVEL": "service_level",
  "SKU": "sku",
  "TOTAL_CAPACITY": "total_capacity",
  "USAGE": "usage",
}
`;

exports[`InventoryCardSubscriptionsContext should handle a store response using selectors: store response 1`] = `
{
  "dataSetColumnHeaders": [
    {
      "cell": [Function],
      "content": "t([curiosity-inventory.header_product_name,curiosity-inventory.guestsHeader_product_name], {"context":"lorem"})",
      "label": [Function],
      "metric": "product_name",
      "width": undefined,
    },
  ],
  "dataSetRows": [],
  "error": false,
  "fulfilled": true,
  "message": null,
  "pending": false,
  "responses": {
    "id": {},
    "list": [
      {
        "data": [
          {
            "data": [
              {
                "product_name": "lorem-ipsum",
              },
              {
                "product_name": "dolor-sit",
              },
            ],
            "meta": {},
          },
        ],
        "fulfilled": true,
      },
    ],
  },
  "resultsColumnCountAndWidths": {
    "count": 1,
    "widths": [
      undefined,
    ],
  },
  "resultsCount": undefined,
  "resultsOffset": undefined,
  "resultsPerPage": undefined,
}
`;

exports[`InventoryCardSubscriptionsContext should handle an onColumnSort event: onColumnSort event, dispatch 1`] = `
[
  [
    [
      {
        "dir": "desc",
        "type": "SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_dir",
        "viewId": "lorem",
      },
      {
        "sort": "loremIpsumColumnOne",
        "type": "SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_sort",
        "viewId": "lorem",
      },
    ],
  ],
  [
    [
      {
        "dir": "asc",
        "type": "SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_dir",
        "viewId": "lorem",
      },
      {
        "sort": "loremIpsumColumnOne",
        "type": "SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_sort",
        "viewId": "lorem",
      },
    ],
  ],
]
`;

exports[`InventoryCardSubscriptionsContext should handle an onPage event: onPage event, dispatch 1`] = `
[
  [
    [
      {
        "offset": 1,
        "type": "SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_offset",
        "viewId": "lorem",
      },
      {
        "limit": 5,
        "type": "SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_limit",
        "viewId": "lorem",
      },
    ],
  ],
]
`;

exports[`InventoryCardSubscriptionsContext should handle variations in instances inventory API responses: inventory, cancelled 1`] = `
{
  "cancelled": true,
}
`;

exports[`InventoryCardSubscriptionsContext should handle variations in instances inventory API responses: inventory, disabled 1`] = `
{
  "data": {},
  "error": false,
  "fulfilled": false,
  "pending": false,
}
`;

exports[`InventoryCardSubscriptionsContext should handle variations in instances inventory API responses: inventory, error 1`] = `
{
  "error": true,
}
`;

exports[`InventoryCardSubscriptionsContext should handle variations in instances inventory API responses: inventory, fulfilled 1`] = `
{
  "fulfilled": true,
}
`;

exports[`InventoryCardSubscriptionsContext should handle variations in instances inventory API responses: inventory, pending 1`] = `
{
  "pending": true,
}
`;

exports[`InventoryCardSubscriptionsContext should parse filters and settings: parsed 1`] = `
{
  "columnCountAndWidths": {
    "count": 1,
    "widths": [
      undefined,
    ],
  },
  "filters": [
    {
      "cell": [Function],
      "header": [Function],
      "label": [Function],
      "metric": "lorem ipsum",
      "width": undefined,
    },
  ],
  "isGuestFiltersDisabled": true,
  "settings": {},
}
`;

exports[`InventoryCardSubscriptionsContext should return specific properties: specific properties 1`] = `
{
  "useGetSubscriptionsInventory": [Function],
  "useInventoryCardActionsSubscriptions": [Function],
  "useOnColumnSortSubscriptions": [Function],
  "useOnPageSubscriptions": [Function],
  "useParseSubscriptionsFiltersSettings": [Function],
  "useSelectorSubscriptions": [Function],
}
`;
